File: fyers_login.py
client_id = 'DO3OHPAO1D-100'
secret_key='6WMPXU4GKD'
redirect_uri = 'https://trade.fyers.in/api-login/redirect-uri/index.html'

from fyers_api import fyersModel,accessToken
from fyers_api.Websocket import ws
import os, time
import pandas as pd
import threading

def get_access_token():
    if not os.path.exists("access_token.txt"):
        session=accessToken.SessionModel(client_id=client_id,secret_key=secret_key,redirect_uri=redirect_uri, response_type='code', grant_type='authorization_code')
        response = session.generate_authcode()
        print("login url: ", response)
        auth_code = input("Enter auth code: ")
        session.set_token(auth_code)
        response = session.generate_token()
        access_token = response["access_token"]
        with open("access_token.txt","w") as f:
            f.write(access_token)
    else:
        with open("access_token.txt","r") as f:
            access_token = f.read()
    return access_token
access_token = get_access_token()
fyers = fyersModel.FyersModel(client_id='DO3OHPAO1D-100', token=access_token)
print(fyers.positions())

def exitAllPositions(profit, loss):
    while True:
        positions = fyers.positions()
        pl_total = fyers.positions()["overall"]["pl_total"]
        data={}
        print(pl_total)
        if pl_total <= loss or pl_total >= profit:
            res = fyers.exit_positions(data)
            print(res["message"])
            break
        time.sleep(1)

def exitIndividualPositions(posId, profit, loss):
    positions = fyers.positions()["netPositions"]
    for x in positions:
        if posId == x['id']:
            data = {'id':x['id']}
            if x['pl'] <= loss or x['pl'] > profit:
                res = fyers.exit_positions(data)
                print(res["message"])

def getOpenPositions():
    positionList=[]
    positions = fyers.positions()["netPositions"]
    for x in positions:
        if x['realized_profit'] != x['pl']:
            positionList.append(x['id'])

    return positionList

profit = float(input("Enter profit you want: "))
loss = float(input("Enter loss you can take: "))
exitAllPositions(profit, -loss)
# threading.Thread(target=exitAllPositions,args=(profit,-loss,)).start()
# exitIndividualPositions()
openPositionList = getOpenPositions()
print(openPositionList)
# exitIndividualPositions('NSE:NIFTY22OCT17800PE-INTRADAY',profit, -loss)

# for pos in openPositionList:
#     threading.Thread(target=exitIndividualPositions,args=(pos,)).start()

File: fyers_login.js
const fyers   = require("fyers-api-v2")
const url = require('url');
const fs = require('fs');

const app_id = "DO3OHPAO1D-100";
const redirect_url = "https://trade.fyers.in/api-login/redirect-uri/index.html";
const key = "6WMPXU4GKD";
var posId ="";

function setAccessTokenForMe(){
    fyers.setAppId(app_id);
    fyers.setRedirectUrl(redirect_url);
    fyers.generateAuthCode();
    const prompt = require("prompt-sync")({ sigint: true });
    //With readline
    const code = prompt("Enter Auth Code");
    const reqBody = {
        auth_code:code,
        secret_key:key
    }

    fyers.generate_access_token(reqBody).then((response)=>{
    console.log(response);
    var token=""+response.access_token;
    fs.writeFile('./accessToken.txt', token, err => {
        if (err) {
            console.error(err);
        }
  // file written successfully
    });

})
}

function exitAllPositions(profit, loss, i){
        fyers.get_positions().then((response) => {
            var pl_total = response.overall.pl_total;
            console.log("P&L at "+i +" seconds = " +pl_total);
            if (pl_total <= loss || pl_total >= profit ){
                const reqBody = {
                     data:{},
                     app_id:app_id,
                     token:accessToken
                }
                const exitPosition=fyers.exit_position(reqBody).then((response) => {
                             console.log(response)
                })
            }

        })
}

function sleep(ms) {
      return new Promise(resolve => setTimeout(resolve, ms));
}

function getOpenPositions(){
    var positionList=[];
    fyers.get_positions().then((response) => {
        var positions = response.netPositions;
        for(let i=0; i<positions.length; i++){
            if (positions[i].realized_profit != positions[i].pl){
                   positionList.push(positions[i].id);
              }
        }
    })
    return positionList;
}


function exitIndividualPositions(posId, profit, loss,i){
    fyers.get_positions().then((response) => {
            var positions = response.netPositions;
            for(let i=0; i<positions.length; i++){
                if(posId == positions[i].id){
                var pl_individual = positions[i].pl;
                console.log("Individual P&L at "+i +" seconds = " +pl_individual);
                    if(pl_individual <= loss || pl_individual >= profit){
                    const reqBody = {
                                      data:{"id":posId},
                                      app_id:app_id,
                                      token:accessToken
                                }
                                const exitPosition = fyers.exit_position(reqBody).then((response) => {
                                                   console.log(response)
                                })
                    }

                }
            }
        })
    }

async function Tutor() {
   for (let i = 1; i <=60*2 ; i++) {
       await sleep(1000);
       exitAllPositions(profit, loss,i);
     //  if(isEmpty(getOpenPositions())){
      //     console.log("All positions are closed: ");
       //    break;
     //  }
//       exitIndividualPositions(posId,profit,loss,i);
   }
}

function isEmpty(obj) {
    return Object.keys(obj).length === 0;
}

//setAccessTokenForMe();

fyers.setAppId(app_id);
fyers.setRedirectUrl(redirect_url);
fs.readFile('./accessToken.txt', 'utf8', (err, data) => {
                       if (err) {
                         console.error(err);
                         return;
                       }
                       fyers.setAccessToken(data);
                     });

const prompt = require("prompt-sync")({ sigint: true });
//With readline
let profit = Number(prompt("Enter profit you want: "));
let loss = 0-Number(prompt("Enter loss you can take: "));
console.log(loss);
Tutor();


